src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java

```diff
@@ -203,12 +215,77 @@ public void onFailure(Throwable e) {
     class ReplicaOperationTransportHandler implements TransportRequestHandler<ReplicaRequest> {
         @Override
         public void messageReceived(final ReplicaRequest request, final TransportChannel channel) throws Exception {
+            new AsyncReplicaAction(request, channel).run();
+        }
+    }
+
+    protected static class RetryOnReplicaException extends IndexShardException {
+
+        public RetryOnReplicaException(ShardId shardId, String msg) {
+            super(shardId, msg);
+        }
+
+        public RetryOnReplicaException(ShardId shardId, String msg, Throwable cause) {
+            super(shardId, msg, cause);
+        }
+    }
+
+    private final class AsyncReplicaAction extends AbstractRunnable {
+        private final ReplicaRequest request;
+        private final TransportChannel channel;
+        // important: we pass null as a timeout as failing a replica is
+        // something we want to avoid at all costs
+        private final ClusterStateObserver observer = new ClusterStateObserver(clusterService, null, logger);
+
+
+        AsyncReplicaAction(ReplicaRequest request, TransportChannel channel) {
+            this.request = request;
+            this.channel = channel;
+        }
+
+        @Override
+        public void onFailure(Throwable t) {
+            if (t instanceof RetryOnReplicaException) {
+                logger.trace("Retrying operation on replica", t);
+                observer.waitForNextChange(new ClusterStateObserver.Listener() {
+                    @Override
+                    public void onNewClusterState(ClusterState state) {
```

Comment:
hmm what threadpool is this executed? I guess we should push it into the indexing threadpool no?